core/src/crds_gossip_pull.rs

```diff
@@ -391,25 +409,27 @@ impl CrdsGossipPull {
     // build a set of filters of the current crds table
     // num_filters - used to increase the likelyhood of a value in crds being added to some filter
     pub fn build_crds_filters(&self, crds: &Crds, bloom_size: usize) -> Vec<CrdsFilter> {
+        const PAR_MIN_LENGTH: usize = 1024;
         let num = cmp::max(
             CRDS_GOSSIP_DEFAULT_BLOOM_ITEMS,
             crds.table.values().count() + self.purged_values.len(),
         );
-        let mut filters = CrdsFilterSet::new(num, bloom_size);
-        let mut add_value_hash = |value_hash| {
-            if let Some(filter) = filters.get(value_hash) {
-                debug_assert!(filter.test_mask(value_hash));
-                filter.filter.add(value_hash);
-            }
-        };
-        for v in crds.table.values() {
-            add_value_hash(&v.value_hash);
-        }
-        for (value_hash, _insert_timestamp) in &self.purged_values {
-            add_value_hash(&value_hash);
-        }
-
-        filters.0
+        let filters = CrdsFilterSet::new(num, bloom_size);
+        rayon::join(
+            || {
```

Comment:
I think using the gossip-specific threadpool here might be a good idea. The thread limit will keep it from using the whole machine's CPU and we've found that not sharing work between other instances of rayon generally performs better.

Reply:
Agreed, but I tried adding a ThreadPool to [`struct CrdsGossipPull`](https://github.com/solana-labs/solana/blob/02ba197a06184e0a6ae2254ad387ee9376fd18c4/core/src/crds_gossip_pull.rs#L160), and it was not compiling with the `#[derive(Clone)]` thing. [`struct CrdsGossip`](https://github.com/solana-labs/solana/blob/02ba197a06184e0a6ae2254ad387ee9376fd18c4/core/src/crds_gossip.rs#L20) also has `
#[derive(Clone)]`. Would it make sense to add the `ThreadPool` to [`struct ClusterInfo`](https://github.com/solana-labs/solana/blob/02ba197a06184e0a6ae2254ad387ee9376fd18c4/core/src/cluster_info.rs#L248) and pass it down the call-stack? or, do you have a different suggestion?

I am not sure those `#[derive(Clone)]` are used anywhere, so we may be able to drop them. Or, alternatively implement `Clone` manually.

Reply:
yea. I was thinking passing it down the call stack. removing `#[derive(Clone)]` or doing `Clone` manually doesn't seem like it would be too bad.

Reply:
added a ThreadPool to `struct ClusterInfo`. please take a look.

Reply:
After adding a dedicated ThreadPool to [`struct ClusterInfo`](https://github.com/solana-labs/solana/blob/afd9bfc45f7c467470d060dac892de878c9976c0/core/src/cluster_info.rs#L248), a number of tests started seg-faulting and crashing, apparently because of going out of memory. so I had to:
* add `#[serial]` to two of the tests in [core/tests/crds_gossip.rs](https://github.com/solana-labs/solana/blob/master/core/tests/crds_gossip.rs).
* limit number of threads in [`clone_with_id`](https://github.com/solana-labs/solana/blob/afd9bfc45f7c467470d060dac892de878c9976c0/core/src/cluster_info.rs#L434) when cloning cluster-info.

Should we also limit the number of threads for the threadpool in cluster-info as well?

Reply:
I see. I think we do have a limit there:
```
 let thread_pool = rayon::ThreadPoolBuilder::new()
     .num_threads(std::cmp::min(get_thread_count(), 8))
     .thread_name(|i| format!("sol-gossip-work-{}", i))
```

Reply:
I was adding an entirely new ThreadPool here. But, from your comment, I realized that I am doing this wrong, because then the gossip service will have two separate thread pools.
At the same time I cannot use the [`"sol-gossip-work-{i}" threads`](https://github.com/solana-labs/solana/blob/afd9bfc45f7c467470d060dac892de878c9976c0/core/src/cluster_info.rs#L2330-L2334), because those are created in [`ClusterInfo::listen` thread](https://github.com/solana-labs/solana/blob/afd9bfc45f7c467470d060dac892de878c9976c0/core/src/gossip_service.rs#L54), whereas this new code is invoked in [`ClusterInfo::gossip` thread](https://github.com/solana-labs/solana/blob/afd9bfc45f7c467470d060dac892de878c9976c0/core/src/gossip_service.rs#L61).

We can reorganize the gossip service code to create the thread pool for all gossip service work in [`GossipService::new`](https://github.com/solana-labs/solana/blob/afd9bfc45f7c467470d060dac892de878c9976c0/core/src/gossip_service.rs#L31), and then share it between [listen and gossip threads](https://github.com/solana-labs/solana/blob/afd9bfc45f7c467470d060dac892de878c9976c0/core/src/gossip_service.rs#L54-L67). It will look like [this commit](https://github.com/behzadnouri/solana/commit/fdd668052dce64f16d442e052c5ee418ab8da9d0). If that looks right to you I will first send that as a separate pull request first.

Reply:
An alternative would be to actually have _two separate thread-pools_, one for [`ClusterInfo::listen` thread](https://github.com/solana-labs/solana/blob/afd9bfc45f7c467470d060dac892de878c9976c0/core/src/gossip_service.rs#L54), and the other for [`ClusterInfo::gossip` thread](https://github.com/solana-labs/solana/blob/afd9bfc45f7c467470d060dac892de878c9976c0/core/src/gossip_service.rs#L61).

It has the advantage that `listen` and `gossip` may not block each other, at the cost of higher memory requirements at max load.

Reply:
Yea. I hesitate to add another thread pool because we already have too many. The pool getting created in gossip service doesn't look too bad and shared between them. The work is not really blocking like an IO-heavy workload might be, so I think it might be fine to share the thread pool.

Reply:
sounds good. I sent out https://github.com/solana-labs/solana/pull/12402

Reply:
Actually, I think the method in this PR is fine where thread_pool is a member of `ClusterInfo` struct. Just move the thread pool creation of gossip-work-{i} to `ClusterInfo::new` and then use `self.thread_pool` in `ClusterInfo::listen`. For test configuration just create a 1 or 2 thread pool.

Reply:
update https://github.com/solana-labs/solana/pull/12402 accordingly